src/video.h

```diff
@@ -11,11 +11,181 @@
+  using sws_t = util::safe_ptr<SwsContext, sws_freeContext>;
+  using img_event_t = std::shared_ptr<safe::event_t<std::shared_ptr<platf::img_t>>>;
+
+  struct encoder_platform_formats_t {
+    virtual ~encoder_platform_formats_t() = default;
+    platf::mem_type_e dev_type;
+    platf::pix_fmt_e pix_fmt_8bit, pix_fmt_10bit;
+  };
+
+  struct encoder_platform_formats_avcodec: encoder_platform_formats_t {
+    using init_buffer_function_t = std::function<util::Either<avcodec_buffer_t, int>(platf::avcodec_encode_device_t *)>;
+
+    encoder_platform_formats_avcodec(
+      const AVHWDeviceType &avcodec_base_dev_type,
+      const AVHWDeviceType &avcodec_derived_dev_type,
+      const AVPixelFormat &avcodec_dev_pix_fmt,
+      const AVPixelFormat &avcodec_pix_fmt_8bit,
+      const AVPixelFormat &avcodec_pix_fmt_10bit,
+      const init_buffer_function_t &init_avcodec_hardware_input_buffer_function):
+        avcodec_base_dev_type { avcodec_base_dev_type },
+        avcodec_derived_dev_type { avcodec_derived_dev_type },
+        avcodec_dev_pix_fmt { avcodec_dev_pix_fmt },
+        avcodec_pix_fmt_8bit { avcodec_pix_fmt_8bit },
+        avcodec_pix_fmt_10bit { avcodec_pix_fmt_10bit },
+        init_avcodec_hardware_input_buffer { init_avcodec_hardware_input_buffer_function } {
+      dev_type = map_base_dev_type(avcodec_base_dev_type);
+      pix_fmt_8bit = map_pix_fmt(avcodec_pix_fmt_8bit);
+      pix_fmt_10bit = map_pix_fmt(avcodec_pix_fmt_10bit);
+    }
+
+    AVHWDeviceType avcodec_base_dev_type, avcodec_derived_dev_type;
+    AVPixelFormat avcodec_dev_pix_fmt;
+    AVPixelFormat avcodec_pix_fmt_8bit, avcodec_pix_fmt_10bit;
+
+    init_buffer_function_t init_avcodec_hardware_input_buffer;
+  };
+
+  struct encoder_platform_formats_nvenc: encoder_platform_formats_t {
+    encoder_platform_formats_nvenc(
+      const platf::mem_type_e &dev_type,
+      const platf::pix_fmt_e &pix_fmt_8bit,
+      const platf::pix_fmt_e &pix_fmt_10bit) {
+      encoder_platform_formats_t::dev_type = dev_type;
+      encoder_platform_formats_t::pix_fmt_8bit = pix_fmt_8bit;
+      encoder_platform_formats_t::pix_fmt_10bit = pix_fmt_10bit;
+    }
+  };
+
+  struct encoder_t {
+    std::string_view name;
+    enum flag_e {
+      PASSED,  // Is supported
+      REF_FRAMES_RESTRICT,  // Set maximum reference frames
+      CBR,  // Some encoders don't support CBR, if not supported --> attempt constant quantatication parameter instead
+      DYNAMIC_RANGE,  // hdr
+      VUI_PARAMETERS,  // AMD encoder with VAAPI doesn't add VUI parameters to SPS
+      MAX_FLAGS
+    };
+
+    static std::string_view
+    from_flag(flag_e flag) {
+#define _CONVERT(x) \
+  case flag_e::x:   \
+    std::string_view(#x)
+      switch (flag) {
+        _CONVERT(PASSED);
+        _CONVERT(REF_FRAMES_RESTRICT);
+        _CONVERT(CBR);
+        _CONVERT(DYNAMIC_RANGE);
+        _CONVERT(VUI_PARAMETERS);
+        _CONVERT(MAX_FLAGS);
+      }
+#undef _CONVERT
+
+      return {"unknown"};
+    }
```

Comment:
This was refactored slightly to avoid `using namespace std::literal;` in the header.